python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,542 @@ def array_to_vector(col: Column) -> Column:
+                            single_input = np.vstack(batch.iloc[:, 0])
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+                    else:
+                        # scalar columns
+                        if len(batch.columns) == 1:
+                            # single scalar column, remove extra dim
+                            single_input = np.squeeze(batch.to_numpy())
+                            if input_shapes and input_shapes[0] not in [None, [], [1]]:
+                                raise ValueError("Invalid input_tensor_shape for scalar column.")
+                        elif not has_tuple:
+                            # columns grouped via struct/array, convert to single tensor
+                            single_input = batch.to_numpy()
+                            if input_shapes and input_shapes[0] != [len(batch.columns)]:
+                                raise ValueError("Input data does not match expected shape.")
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+
+                    # if input_tensor_shapes provided, try to reshape input
+                    if input_shapes:
+                        if len(input_shapes) == 1:
+                            single_input = single_input.reshape(
+                                [-1] + input_shapes[0]  # type: ignore
+                            )
+                            if len(single_input) != len(batch):
+                                raise ValueError("Input data does not match expected shape.")
+                        else:
+                            raise ValueError(
+                                "Multiple input_tensor_shapes found, but model expected one input"
+                            )
+
+                    # run model prediction function on transformed (numpy) inputs
+                    preds = predict_fn(single_input)
+                else:
+                    msg = "Model expected {} inputs, but received {} columns"
+                    raise ValueError(msg.format(num_expected_cols, num_input_cols))
+
+                # return predictions to Spark
+                _validate(preds, num_input_rows, return_type)
+                if isinstance(return_type, StructType):
+                    yield pd.DataFrame(preds)
+                elif isinstance(return_type, ArrayType):
+                    yield pd.Series(list(preds))  # type: ignore[misc]
+                else:
+                    yield pd.Series(np.squeeze(preds))  # type: ignore
```

Comment:
I suggest to merge the part code logic with  `_validate` function, and rename the merged function as `_validate_and_generate_returned_value`